source/MaterialXGenShader/Util.cpp

```diff
@@ -13,119 +13,126 @@ namespace MaterialX
+    bool isEqual(ValuePtr value, const float& value2)
     {
-        if (value->isA<float>() && isZero(value->asA<float>()))
+        if (value->isA<float>() && isEqual(value->asA<float>(), value2))
         {
             return true;
         }
-        if (value->isA<Color3>() && isZero(value->asA<Color3>()))
+        else if (value->isA<Color3>())
         {
-            return true;
+            const Color3& value3 = value->asA<Color3>();
+            if (isEqual(value3[0], value2) && isEqual(value3[1], value2) && isEqual(value3[2], value2))
+            {
+                return true;
+            }
         }
         return false;
     }
 
-    bool isOne(float v)
-    {
-        return v > EPS_ONE;
-    }
+    using OpaqueTestPair = std::pair<string, float>;
+    using OpaqueTestPairList = vector<OpaqueTestPair>;
 
-    bool isOne(const Color3& c)
+    InputPtr getInputInterface(const string& interfaceName , NodePtr node)
     {
-        return isOne(c[0]) && isOne(c[1]) && isOne(c[2]);
+        InputPtr interfaceInput = nullptr;
+        ElementPtr parent = node->getParent();
+        NodeGraphPtr nodeGraph = parent ? parent->asA<NodeGraph>() : nullptr;
+        if (nodeGraph)
+        {
+            interfaceInput = nodeGraph->getInput(interfaceName);
+            if (!interfaceInput)
+            {
+                NodeDefPtr nodeDef = nodeGraph->getNodeDef();
+                if (nodeDef)
+                {
+                    interfaceInput = nodeDef->getInput(interfaceName);
+                }
+            }
+        }
+        return interfaceInput;
     }
 
-    bool isOne(ValuePtr value)
+    bool hasTransparentInputs(const OpaqueTestPairList& opaqueInputList, NodePtr node)
     {
-        if (value->isA<float>() && isOne(value->asA<float>()))
-        {
-            return true;
-        }
-        if (value->isA<Color3>() && isOne(value->asA<Color3>()))
+        for (auto opaqueInput : opaqueInputList)
         {
-            return true;
+            InputPtr interfaceInput = node->getInput(opaqueInput.first);
+            if (interfaceInput)
+            {
+                if (interfaceInput->getConnectedNode())
+                {
+                    return true;
+                }
+                ValuePtr value = interfaceInput->getValue();
+                if (value && !isEqual(value, opaqueInput.second))
+                {
+                    return true;
+                }
+            }
         }
         return false;
     }
 
-    bool isTransparentShaderNode(NodePtr node)
+    bool isTransparentShaderNode(NodePtr node, OpaqueTestPairList& interfaceNames)
     {
         if (!node || node->getType() != SURFACE_SHADER_TYPE_STRING)
         {
             return false;
         }
 
-        // Check opacity
-        InputPtr opacity = node->getActiveInput("opacity");
-        if (opacity)
+        // Current set if inputs and values which indicate opacity
+        OpaqueTestPairList inputPairList;
```

Comment:
Nitpick: Using an initializer list might be more efficient than calling push_back multiple times (since push_back might cause the vector to realloc to a larger size, and in this case we already know the size).

Reply:
I'll clean this up. At some point it would be good to have this dynamic if we support tagging inputs as "affects transparency".

Reply:
Sure, and you can still have the list as a dynamic std::vector, I was just nitpicky and pointing out that these first three push_back calls could be done with an initializer list instead.